site/sfguides/src/production_ready_rag_with_trulens_snowflake_cortex/production_ready_rag_trulens_snowflake_cortex.md

```diff
@@ -0,0 +1,483 @@
+    [".md"],
+    GithubRepositoryReader.FilterType.INCLUDE,
+  )
+)
+
+documents = reader.load_data(branch="main")
+
+import re
+
+def clean_up_text(content: str) -> str:
+  """
+  Remove unwanted characters and patterns in text input.
+
+  :param content: Text input.
+  
+  :return: Cleaned version of original text input.
+  """
+
+  # Fix hyphenated words broken by newline
+  content = re.sub(r'(\w+)-\n(\w+)', r'\1\2', content)
+
+  unwanted_patterns = ['---\nvisible: false','---', '#','slug:']
+  for pattern in unwanted_patterns:
+    content = re.sub(pattern, "", content)
+
+  # Remove all slugs starting with a \ and stopping at the first space
+  content = re.sub(r'\\slug: [^\s]*', '', content)
+
+  # normalize whitespace
+  content = re.sub(r'\s+', ' ', content)
+  return content
+
+cleaned_documents = []
+
+for d in documents:
+  cleaned_text = clean_up_text(d.text)
+  d.text = cleaned_text
+  cleaned_documents.append(d)
+```
+
+### Embed the preprocessed documents
+
+We'll use Snowflake's Arctic Embed model available from HuggingFace to embed the documents. We'll also use Llama-Index's `SemanticSplitterNodeParser` for processing.
+
+```python
+from llama_index.embeddings.huggingface import HuggingFaceEmbedding
+from llama_index.core.node_parser import SemanticSplitterNodeParser
+
+embed_model = HuggingFaceEmbedding("Snowflake/snowflake-arctic-embed-m")
+
+splitter = SemanticSplitterNodeParser(
+  buffer_size=1, breakpoint_percentile_threshold=85, embed_model=embed_model
+)
+```
+
+With the embed model and splitter, we can execute them in an ingestion pipeline
+
+```python
+from llama_index.core.ingestion import IngestionPipeline
+
+cortex_search_pipeline = IngestionPipeline(
+  transformations=[
+    splitter,
+  ],
+)
+
+results = cortex_search_pipeline.run(show_progress=True, documents=cleaned_documents)
+```
+
+### Load data to Cortex Search
+
+Now that we've embedded our documents, we're ready to load them to Cortex Search.
+
+Here we can use the same connection details as we set up for Cortex Complete.
+
+```python
+import os
+import snowflake.connector
+from tqdm.auto import tqdm
+
+conn = snowflake.connector.connect(
+  user=connection_details["user"],
+  password=connection_details["password"],
+  account=connection_details["account"],
+  warehouse=connection_details["warehouse"],
+  database=connection_details["database"],
+  schema=connection_details["schema"]
+)
+
+conn.cursor().execute("CREATE OR REPLACE TABLE streamlit_docs(doc_text VARCHAR)")
+for curr in tqdm(result):
+  conn.cursor().execute("INSERT INTO streamlit_docs VALUES (%s)", curr.text)
+```
+
+## Calling the Cortex Search Service
+Duration: 5
+
+Here we'll create a CortexSearchRetreiver class to connect to our cortex search service and add the `retrieve` method that we can leverage for calling it.
+
+```python
+import os
+from snowflake.core import Root
+
+class CortexSearchRetriever:
+
+  def __init__(self, session = session, limit_to_retrieve: int = 4):
```

Comment:
why do you have the default arg for `session`? It's not necessary given you're not using it and kinda confusing since they're named the same (in fact, there's confusion down below on L251). Also it's not typed.